src/mlir/lowering.md

```diff
@@ -0,0 +1,517 @@
+
+The syntax is similar to the static call:
+```
+  %0 = something that gives an runtime object of class Foo
+  %res = verona.call "method" [%0] (%arg0, %arg1...)
+```
+
+With `method` being a field name, not an actual function/method. Depending on
+how hard it is to discern this at compile time, we may want to change the syntax
+of the operation (or create a new one) to make it clear it's a field, not a
+method.
+
+Lambdas are anonymous structures already initialised by the declaration, with
+an `apply` method, so the call mechanism is exactly the same.
+
+### WhenStmt
+
+Scheduled behaviour as a lambda initialised with the captures as usual. The
+`apply` arguments are the `cown`s in the `when` clause. In additional to the
+lambda, a `when` statement also calls the runtime library to schedule the
+behaviour (push it to the work queue).
+
+So, a `WhenStmt` is roughly lowered to the sequence:
+```
+  std.call Anon$123.create(capture1, capture2,...);
+  %descriptor = verona.static(Anon$123);
+  std.call rt.Cown.schedule(%descriptor, %cown1, %cown2, ...);
+```
+
+### ReadFieldStmt / WriteFieldStmt
+
+Access class fields for read/write. The location of the field and the operations
+to extract them are defined in the ABI, which hasn't been calculated yet, so
+we need dialect operations to keep the information on the field name until the
+ABI pass converts field access into lowering patterns.
+
+The operation is similar to method calls:
+```
+  %0 = something that gives an runtime object of class Foo
+
+  // Read returns the field value
+  %res = verona.field_read "field" [%0]
+
+  // Write returns the previous field value
+  %new_val = constant 10
+  %old_val = verona.field_read "field" [%new_val]
+```
+
+There are different rules for accessing single typed fields and union types. The
+current ABI proposal lays out in order of size, from largest to smallest, and
+therefore the offset from the object's address is not the same as the sum of all
+previously declared (in program order).
+
+Furthermore, union types may have compacting optimisations and discriminators to
+match which value is stored, so will need non-trivial extraction lowering (mask
+and shift on the discriminator, then mask, shift and possibly extend on the
+payload). Writes will have to do the reverse process to save values in the ABI
+representation.
+
+### <Integer|Float>LiteralStmt
+
+Constant literals with generic type (int, float). The type we assign to them
+will depend on how they're used and therefore will depend on type inference and
+reachability analysis, which should have been completed by the time the 1IR is
+ready to be converted in the 2IR.
+
+Literals are lowered as immutable objects of the specified concrete type,
+initialised with the literal value.
+
+Example:
+```
+  // 10 + a : U32
+  %const = verona.constant "10" : U32
+  %0 = verona.new_object @U32 [ %const ]
+  %1 = verona.call "+" [ %0 ] ( %a )
+```
+
+### StringLiteralStmt
```

Comment:
Can we just ignore strings for now? It's a huge can of worms. I wrote the current implementation just for the sake of demonstration purposes, so I wouldn't build anything on top of it for now.

I suspect 80% of the functionality of strings will be managed in the standard library and the rest will almost entirely be in the frontend. Anything that's left in the backend will likely be trivial and actually be more about array literals than anything else.